src/client/StripeReader.cpp

```diff
@@ -0,0 +1,453 @@
+                           threadPool(ThreadPool::getInstance(
+                                   min((long) conf->getStripeReaderThreadPoolSize(),
+                                       sysconf(_SC_NPROCESSORS_ONLN) * 2))) {
+    this->dataBlkNum = ecPolicy->getNumDataUnits();
+    this->parityBlkNum = ecPolicy->getNumParityUnits();
+    this->cellSize = ecPolicy->getCellSize();
+}
+
+StripeReader::~StripeReader() {
+}
+
+void StripeReader::updateState4SuccessRead(StripedBlockUtil::StripingChunkReadResult & result) {
+    Preconditions::checkArgument(
+        result.state == StripedBlockUtil::StripingChunkReadResult::SUCCESSFUL);
+
+    readerInfos[result.index]->setOffset(alignedStripe.getOffsetInBlock() + alignedStripe.getSpanInBlock());
+}
+
+void StripeReader::checkMissingBlocks() {
+    if (alignedStripe.missingChunksNum > parityBlkNum) {
+        clearFutures();
+        THROW(HdfsIOException, "Missing blocks, missingChunksNum: %d, parityBlkNum: %d",
+              alignedStripe.missingChunksNum, parityBlkNum);
+    }
+}
+
+/**
+ * We need decoding. Thus go through all the data chunks and make sure we
+ * submit read requests for all of them.
+ */
+void StripeReader::readDataForDecoding() {
+    prepareDecodeInputs();
+    for (int i = 0; i < dataBlkNum; i++) {
+        Preconditions::checkNotNull(alignedStripe.chunks[i]);
+        if (alignedStripe.chunks[i]->state == StripedBlockUtil::StripingChunk::REQUESTED) {
+            if (!readChunk(targetBlocks[i], i)) {
+                alignedStripe.missingChunksNum++;
+            }
+        }
+    }
+    checkMissingBlocks();
+}
+
+void StripeReader::readParityChunks(int num) {
+    for (int i = dataBlkNum, j = 0; i < dataBlkNum + parityBlkNum && j < num; i++) {
+        if (alignedStripe.chunks[i] == nullptr) {
+            if (prepareParityChunk(i) && readChunk(targetBlocks[i], i)) {
+                j++;
+            } else {
+                alignedStripe.missingChunksNum++;
+            }
+        }
+    }
+    checkMissingBlocks();
+}
+
+std::vector<shared_ptr<ReaderStrategy>> StripeReader::getReadStrategies(StripedBlockUtil::StripingChunk & chunk) {
+    std::vector<shared_ptr<ReaderStrategy>> strategies;
+    shared_ptr<ReaderStrategy> strategy = shared_ptr<ReaderStrategy>(new ReaderStrategy(*chunk.getByteBuffer()));
+    strategies.push_back(strategy);
+    return strategies;
+}
+
+int32_t StripeReader::readToBuffer(const LocatedBlock & curBlock,
+                                   shared_ptr<BlockReader> blockReader,
+                                   shared_ptr<ReaderStrategy> strategy,
+                                   const DatanodeInfo & currentNode) {
+    int32_t length = 0;
+    try {
+        while (length < strategy->targetLength) {
+            int32_t ret = readFromBlock(curBlock, currentNode, blockReader,
+                                        (char *)strategy->readBuf.getBuffer() + strategy->readBuf.position(),
```

Comment:
static_cast

Reply:
static_cast from 'int8_t *' (aka 'signed char *') to 'char *' is not allowed.